sdk/identity/azure_identity/src/managed_identity_credential.rs

```diff
@@ -256,32 +220,35 @@ mod tests {
         });
         let mut options = options.unwrap_or_default();
         options.credential_options = TokenCredentialOptions {
+            env,
             http_client: Arc::new(mock_client),
             ..Default::default()
         };
         let cred = ManagedIdentityCredential::new(Some(options)).expect("credential");
-        let token = cred.get_token(LIVE_TEST_SCOPES).await.expect("token");
-        assert_eq!(token.expires_on.unix_timestamp(), expires_on as i64);
-        assert_eq!(token.token.secret(), "*");
+        for _ in 0..4 {
+            let token = cred.get_token(LIVE_TEST_SCOPES).await.expect("token");
+            assert_eq!(token.expires_on.unix_timestamp(), expires_on as i64);
+            assert_eq!(token.token.secret(), "*");
+            assert_eq!(token_requests.load(Ordering::SeqCst), 1);
+        }
     }
 
-    fn run_unsupported_source_test(
-        env_vars: HashMap<&str, &str>,
-        expected_source: ManagedIdentitySource,
-    ) {
-        let _guard = EnvVarGuard::new(&env_vars);
-        let actual_source = get_source();
-        assert!(
-            std::mem::discriminant(&actual_source) == std::mem::discriminant(&expected_source),
-            "Expected {:?}, got {:?}",
-            expected_source,
-            actual_source
+    fn run_unsupported_source_test(env: Env, expected_source: ManagedIdentitySource) {
+        let actual_source = get_source(&env);
+        assert_eq!(
+            std::mem::discriminant(&actual_source),
+            std::mem::discriminant(&expected_source)
         );
-        let result = ManagedIdentityCredential::new(None);
+        let result = ManagedIdentityCredential::new(Some(ManagedIdentityCredentialOptions {
+            credential_options: TokenCredentialOptions {
+                env,
+                ..Default::default()
+            },
+            ..Default::default()
+        }));
         assert!(
             matches!(result, Err(ref e) if *e.kind() == azure_core::error::ErrorKind::Credential),
-            "Expected constructor error, got: {:?}",
-            result
+            "Expected constructor error"
```

Comment:
Perhaps not in this PR, but generally in Rust, "if it compiles it works". If we're asserting something at runtime that we could otherwise do at compile time e.g., only certain variants are allowed, certain strings are valid, whatnot, that should be modeled as such in code. Obviously some things are out of our control e.g., bad content to deserialize.

Reply:
I also prefer build to runtime errors but I don't see a way to shift this one left because we can't know before runtime which managed identity API(s) an application will actually use. For example, keeping `VirtualMachineManagedIdentityCredential` and friends in the public API would force applications to make their expectations explicit but we'd still have to return an error at runtime when the type didn't match the host